identity/manager.go

```diff
@@ -110,34 +110,126 @@ func (m *Manager) Create(ctx context.Context, i *Identity, opts ...ManagerOption
+				continue
+			}
+
+			// FindByCredentialsIdentifier does not expand identity credentials.
+			if err = m.r.PrivilegedIdentityPool().HydrateIdentityAssociations(ctx, found, ExpandCredentials); err != nil {
+				return err
+			}
 		}
-		if err != nil {
-			m.r.Logger().WithError(err).Error("Failed to generate credential hint")
-			continue
+	}
+
+	// If the conflict is not in the identifiers table, it is coming from the verifiable or recovery address.
+	var foundConflictAddress string
+	if found == nil {
+		for _, va := range i.VerifiableAddresses {
+			conflictingAddress, err := m.r.PrivilegedIdentityPool().FindVerifiableAddressByValue(ctx, va.Via, va.Value)
+			if errors.Is(err, sqlcon.ErrNoRows) {
+				continue
+			} else if err != nil {
+				return err
+			}
+
+			foundConflictAddress = conflictingAddress.Value
+			found, err = m.r.PrivilegedIdentityPool().GetIdentity(ctx, conflictingAddress.IdentityID, ExpandCredentials)
+			if err != nil {
+				return err
+			}
 		}
-		message = fmt.Sprintf("An account for %s already exists.", va.Value)
-		hintIdentity, err := m.r.PrivilegedIdentityPool().GetIdentity(ctx, conflictingAddress.IdentityID, ExpandCredentials)
-		if err != nil {
-			m.r.Logger().WithError(err).Error("Failed to generate credential hint")
+	}
+
+	// Last option: check the recovery address
+	if found == nil {
+		for _, va := range i.RecoveryAddresses {
+			conflictingAddress, err := m.r.PrivilegedIdentityPool().FindRecoveryAddressByValue(ctx, va.Via, va.Value)
+			if errors.Is(err, sqlcon.ErrNoRows) {
+				continue
+			} else if err != nil {
+				return err
+			}
+
+			foundConflictAddress = conflictingAddress.Value
+			found, err = m.r.PrivilegedIdentityPool().GetIdentity(ctx, conflictingAddress.IdentityID, ExpandCredentials)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Still not found? Return generic error.
+	if found == nil {
+		return &ErrDuplicateCredentials{error: e}
+	}
+
+	for _, cred := range found.Credentials {
+		if cred.Config == nil {
 			continue
 		}
-		for _, c := range hintIdentity.Credentials {
-			hints = append(hints, toPretty(&c)...)
+
+		// Basically, we only have password, oidc, and webauthn as first factor credentials.
+		// We don't care about second factor, because they don't help the user understand how to sign
+		// in to the first factor (obviously).
+		switch cred.Type {
+		case CredentialsTypePassword:
+			identifierHint := foundConflictAddress
+			if len(cred.Identifiers) > 0 {
+				identifierHint = cred.Identifiers[0]
+			}
+			return &ErrDuplicateCredentials{
+				error:                e,
+				availableCredentials: []CredentialsType{cred.Type},
+				identifierHint:       identifierHint,
+			}
+		case CredentialsTypeOIDC:
+			var cfg CredentialsOIDC
+			if err := json.Unmarshal(cred.Config, &cfg); err != nil {
+				return errors.WithStack(herodot.ErrInternalServerError.WithReasonf("Unable to JSON decode identity credentials %s for identity %s.", cred.Type, found.ID))
```

Comment:
Since this is for a hint only, IMO we should degrade gracefully here. That is, log the error and continue.

Reply:
Hm but this state should never happen (tm) so it's good if we see an alert for this? It says that the data is corrupt